x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/local/EnableSpatialDistancePushdown.java

```diff
@@ -44,23 +74,125 @@ public class EnableSpatialDistancePushdown extends PhysicalOptimizerRules.Parame
+     * one that needs to be kept after the EVAL.
+     * For example:
+     * <pre>
+     *     FROM index
+     *     | EVAL distance = ST_DISTANCE(field, TO_GEOPOINT("POINT(0 0)")), other = scale * 2
+     *     | WHERE distance &lt;= 10000 AND distance &gt; 5000 AND other &gt; 10
+     * </pre>
+     * Should be rewritten as:
+     * <pre>
+     *     FROM index
+     *     | WHERE ST_INTERSECTS(field, TO_GEOSHAPE("CIRCLE(0,0,10000)"))
+     *         AND ST_DISJOINT(field, TO_GEOSHAPE("CIRCLE(0,0,5000)"))
+     *     | EVAL distance = ST_DISTANCE(field, TO_GEOPOINT("POINT(0 0)")), other = scale * 2
+     *     | WHERE other &gt; 10
+     * </pre>
+     */
+    private PhysicalPlan rewriteBySplittingFilter(FilterExec filterExec, EvalExec evalExec, EsQueryExec esQueryExec) {
+        // Find all pushable distance functions in the EVAL
+        Map<NameId, StDistance> distances = getPushableDistances(evalExec.fields());
+
+        // Don't do anything if there are no distances to push down
+        if (distances.isEmpty()) {
+            return filterExec;
+        }
+
+        // Process the EVAL to get all aliases that might be needed in the filter rewrite
+        AttributeMap<Attribute> aliasReplacedBy = getAliasReplacedBy(evalExec);
+
+        // First we split the filter into multiple AND'd expressions, and then we evaluate each individually for distance rewrites
+        List<Expression> pushable = new ArrayList<>();
+        List<Expression> nonPushable = new ArrayList<>();
+        for (Expression exp : splitAnd(filterExec.condition())) {
+            Expression resExp = exp.transformUp(ReferenceAttribute.class, r -> aliasReplacedBy.resolve(r, r));
             // Find and rewrite any binary comparisons that involve a distance function and a literal
-            var rewritten = filterExec.condition().transformDown(EsqlBinaryComparison.class, comparison -> {
-                ComparisonType comparisonType = ComparisonType.from(comparison.getFunctionType());
-                if (comparison.left() instanceof StDistance dist && comparison.right().foldable()) {
-                    return rewriteComparison(comparison, dist, comparison.right(), comparisonType);
-                } else if (comparison.right() instanceof StDistance dist && comparison.left().foldable()) {
-                    return rewriteComparison(comparison, dist, comparison.left(), ComparisonType.invert(comparisonType));
-                }
-                return comparison;
-            });
-            if (rewritten.equals(filterExec.condition()) == false) {
-                plan = new FilterExec(filterExec.source(), filterExec.child(), rewritten);
+            var rewritten = rewriteDistanceFilters(resExp, distances);
+            // If all pushable StDistance functions were found and re-written, we need to re-write the FILTER/EVAL combination
+            if (rewritten.equals(resExp) == false && conditionHasNoReferences(rewritten, distances)) {
```

Comment:
Now that I thought a bit more about it, I think this condition is too coarse. In particular, we are pushing down the distance computation even if we cannot push down to Lucene in some cases:
```
FROM test | eval dist = st_distance(x, \"POINT(0 0)\"::geo_point) | where dist > 5 OR x::keyword == \"asdf\"

[2024-10-10T15:02:59,037][INFO ][o.e.x.e.o.LocalPhysicalPlanOptimizer] [runTask-0] Rule local.EnableSpatialDistancePushdown applied
ProjectExec[[x{f}#14, y{f}#15, dist{r}#11]]                                                                 = ProjectExec[[x{f}#14, y{f}#15, dist{r}#11]]
\_LimitExec[1000[INTEGER]]                                                                                  = \_LimitExec[1000[INTEGER]]
  \_FilterExec[dist{r}#11 > 5[INTEGER] OR TOSTRING(x{f}#14) == [61 73 64 66][KEYWORD]]                      !   \_EvalExec[[STDISTANCE(x{f}#14,[1 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0][GEO_POINT]) AS dist]]
    \_EvalExec[[STDISTANCE(x{f}#14,[1 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0][GEO_POINT]) AS dist]]        !     \_FilterExec[SPATIALDISJOINT(x{f}#14,[1 11 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 14 40][GEO_SHAPE]) OR TOSTRING
      \_EsQueryExec[test], indexMode[standard], query[][_doc{f}#16], limit[], sort[] estimatedRowSize[null] ! (x{f}#14) == [61 73 64 66][KEYWORD]]
                                                                                                            !       \_EsQueryExec[test], indexMode[standard], query[][_doc{f}#16], limit[], sort[] estimatedRowSize[null]
```
There was no subsequent run of `PushFiltersToSource` that would've done anything about the filter - so we made the query worse by computing the distance twice!

Reply:
Since the issue was that we re-wrote even if the next PushFiltersToSource would fail to push the rewritten filter, the general condition here was to use `canPushToSource` on the rewritten filter, which returns false if it contains any component that blocks the pushdown. So this is now fixed, and there is a text verifying it.

The old `conditionHasNotReferences` is deleted, as it was too specific, as you said.